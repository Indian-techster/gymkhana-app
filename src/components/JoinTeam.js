import React from 'react'
import { connect } from 'react-redux'
import { setTeam, setTeamId } from '../js/actions/index'
import { Redirect } from 'react-router-dom'
import axios from 'axios'
import Form from 'react-bootstrap/Form'
import { FaExclamationCircle} from 'react-icons/fa/'
import Button from 'react-bootstrap/Button'
import Alert from 'react-bootstrap/Alert'
import Loading from '../components/Loading'
import { SERVER_ENDPOINT  } from '../api-config'

/** Redux function. Sirve para enviar (dispatch) acciones al store */
function mapDispatchToProps(dispatch) {
    return {
        setTeam: team => dispatch(setTeam(team)),
        setTeamId: teamId => dispatch(setTeamId(teamId)),
    }
}

const mapStateToProps = state => {
    return { 
        game: state.game,
        username: state.username,
        userid: state.userid,
        team: state.team
    }
}

class Inicio extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            loading: false,
            error: null,
            invalidKeyTeam: false,
            teamKey: "",
            redirect: false
        }
        this.userId = this.props.userid
        this.username = this.props.username
        this.gameId = this.props.game
        this.handleChange = this.handleChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }

    handleChange(event) {
        const value = event.target.value
        const name = event.target.name
        this.setState({
          [name]: value,
          invalidKeyTeam: false,
          emptyInput: false,
          error: null,
        });
    }

    handleSubmit(e){
        e.preventDefault()
        const teamKey = this.state.teamKey
        if(!teamKey) {
            this.setState({ emptyInput: true })
        }
        // else if (teamKey === "test") {
        //     team = "test"
        // }
        else {
            this.setState({ loading: true })
            axios.post(`${SERVER_ENDPOINT}/joinTeam`, {userId: this.userId, key: teamKey, gameId: this.gameId })
            .then(res => {
                this.setState({ loading: false, error: false })
                if(res.data.error) {
                    this.setState({ error: res.data.error })
                }
                else if(!res.data.valid) {
                    this.setState({ invalidKeyTeam: true })
                }
                else if(res.data.result) {
                    debugger
                    const teamName = res.data.result.name
                    const teamId = res.data.result.id
                    this.props.setTeam(teamName)
                    this.props.setTeamId(teamId)
                    
                    this.setState({ redirect: true })
                }
            })
            .catch(error => this.setState({ loading: false, error: error.message })) 
        }
    }
    render() {
        const inputError = (this.state.invalidKeyTeam || this.state.emptyInput) ? true : false
        const inputErrorClassName = inputError ? "g-input-error" : ""

        if(this.state.redirect) {
            return <Redirect to='/intro' />
        }
        return  <React.Fragment>
                {this.state.loading && <Loading/>}
                {this.state.error && <Alert variant="danger">Error: {this.state.error}</Alert>}
                <p className="g-welcome-message">Â¡Hola {this.username}!</p>
                <p className="g-message">{this.props.message}</p>
                <Form onSubmit={this.handleSubmit}>
                    <Form.Group controlId="formBasicEmail">
                        <Form.Control className={"g-input "+ inputErrorClassName} type="text" placeholder="Clave de tu equipo" name="teamKey" value={this.state.teamKey} onChange={this.handleChange} />
                        { this.state.invalidKeyTeam && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ups! Esta clave no existe</Form.Text>}
                        { this.state.emptyInput && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ey! Necesitas indicar la clave</Form.Text>}
                        </Form.Group>
                    <Button className="g-btn" variant="primary" type="submit">OK, go!</Button>
                </Form>
        </React.Fragment>
    }
}

const inicioConnected = connect(mapStateToProps, mapDispatchToProps)(Inicio)
export default inicioConnected;
