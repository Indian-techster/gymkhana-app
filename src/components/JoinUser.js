import React from 'react'
import { connect } from 'react-redux'
import { setUsername, setUserId, restartPoints } from '../js/actions/index'
import axios from 'axios'
import Form from 'react-bootstrap/Form'
import { FaExclamationCircle} from 'react-icons/fa/'
import Button from 'react-bootstrap/Button'
import Alert from 'react-bootstrap/Alert'
import Loading from '../components/Loading'
import { SERVER_ENDPOINT  } from '../api-config'

/** Redux function. Sirve para enviar (dispatch) acciones al store */
function mapDispatchToProps(dispatch) {
    return {
        setUsername: username => dispatch(setUsername(username)),
        setUserId: userid => dispatch(setUserId(userid)),
        restartPoints: points => dispatch(restartPoints())
    }
}

const mapStateToProps = state => {
    return { 
        game: state.game,
        username: state.username,
        // team: state.team
    }
}

class Inicio extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            loading: false,
            error: null,
            duplicatedName: false,
            username: ""
        }
        this.handleChange = this.handleChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)

        // this.textoIntro = "Estás a punto de empezar la aventura más emocionante que ha ocurrido nunca en esta ciudad. "+
       // "Un tesoro se esconde en estas tierras. Para poder encontrarlo, antes debéis demostrar que sois verdaderos piratas."
        //"Solo el equipo que consiga demostrar mayor liderazgo, creatividad e ingenio, será digno de merecer este puesto al más valiente. "
    }
    
    handleChange(event) {
        const value = event.target.value
        const name = event.target.name
        this.setState({
          [name]: value,
          duplicatedName: false,
          emptyInput: false,
          error: null,
        });
    }

    handleSubmit(e){
        e.preventDefault()
        this.props.restartPoints()
        const username = this.state.username
        if(!username) {
            this.setState({ emptyInput: true })
        }
        else {
            this.setState({ loading: true })
            axios.post(`${SERVER_ENDPOINT}/joinUser`, {username: username, game_id: this.props.game })
            .then(res => {
                this.setState({ loading: false, error: false })
                if(res.data.duplicated) {
                    this.setState({ duplicatedName: true })
                }
                else {
                    this.props.setUserId(res.data.result.id)
                    this.props.setUsername(username) //esto provocará que se actualice el state y se renderice la página co joinTeam
                }
            })
            .catch(error => this.setState({ loading: false, error: error.message })) 
        }
    }

    render() {
        const inputError = (this.state.duplicatedName || this.state.emptyInput) ? true : false
        const inputErrorClassName = inputError ? "g-input-error" : ""

        return  <React.Fragment>
            {this.state.loading && <Loading/>}
            {this.state.error && <Alert variant="danger">Error: {this.state.error}</Alert>}
            <p className="g-welcome-message">{this.props.welcomeMessage}</p>
            <Form onSubmit={this.handleSubmit}>
                <Form.Group controlId="formBasicEmail">
                    <Form.Control className={"g-input "+ inputErrorClassName} type="text" placeholder="Nickname" name="username" value={this.state.username} onChange={this.handleChange} />
                    { this.state.duplicatedName && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ups! Este nombre ya existe</Form.Text>}
                    { this.state.emptyInput && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ey! No olvides poner tu nombre</Form.Text>}
                </Form.Group>
                <Button className="g-btn" variant="primary" type="submit">Siguiente</Button>
            </Form>
        </React.Fragment>
    }
}

const inicioConnected = connect(mapStateToProps, mapDispatchToProps)(Inicio)
export default inicioConnected;
