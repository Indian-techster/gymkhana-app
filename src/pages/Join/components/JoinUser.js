import React from 'react'
import axios from 'axios'
import Form from 'react-bootstrap/Form'
import { FaExclamationCircle} from 'react-icons/fa/'
import Button from 'react-bootstrap/Button'
import Alert from 'react-bootstrap/Alert'
import Loading from 'components/Loading/Loading'
import { SERVER_ENDPOINT  } from 'api-config'


class JoinUser extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            loading: false,
            error: null,
            duplicatedName: false,
            username: ""
        }
        this.handleChange = this.handleChange.bind(this)
        this.handleSubmit = this.handleSubmit.bind(this)
    }
    
    handleChange(event) {
        const value = event.target.value
        const name = event.target.name
        this.setState({
          [name]: value,
          duplicatedName: false,
          emptyInput: false,
          error: null,
        })
    }

    handleSubmit(e){
        e.preventDefault()
        const username = this.state.username
        if(!username) {
            this.setState({ emptyInput: true })
        }
        else {
            this.setState({ loading: true })
            axios.post(`${SERVER_ENDPOINT}/game/joinUser`, {username: username, gameId: this.props.gameid })
            .then(res => {
                this.setState({ loading: false, error: false })
                if(res.data.duplicated) {
                    this.setState({ duplicatedName: true })
                }
                else if(res.data.error) {
                    this.setState({ error: res.data.error.detail })
                }
                else {
                    this.props.setUser(res.data.result.id, username) 
                }
            })
            .catch(error => this.setState({ loading: false, error: error.message })) 
        }
    }

    render() {
        const inputError = (this.state.duplicatedName || this.state.emptyInput) ? true : false
        const inputErrorClassName = inputError ? "g-input-error" : ""

        return (
            <React.Fragment>
                {this.state.loading && <Loading/>}
                {this.state.error && <Alert variant="danger">Error JoinUser: {this.state.error}</Alert>}
                <p className="join-welcome-message">{this.props.welcomeMessage}</p>
                <Form onSubmit={this.handleSubmit}>
                    <Form.Group controlId="formBasicEmail">
                        <Form.Control className={"g-input "+ inputErrorClassName} type="text" placeholder="Nickname" name="username" value={this.state.username} onChange={this.handleChange} />
                        { this.state.duplicatedName && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ups! Este nombre ya existe</Form.Text>}
                        { this.state.emptyInput && <Form.Text className="g-invalid-input-warning"><FaExclamationCircle/> Ey! No olvides poner tu nombre</Form.Text>}
                    </Form.Group>
                    <Button className="g-btn" variant="primary" type="submit">Siguiente</Button>
                </Form>
            </React.Fragment>
        )
    }
}

export default JoinUser
