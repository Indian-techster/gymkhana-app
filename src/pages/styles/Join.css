@keyframes slideInFromTop {
    0% {
      /* transform: translateY(-100%); */
      transform: translateY(calc(-50vh));
    }
    100% {
      transform: translateY(0);
    }
}

.join-welcome-message {
  top: 20%;
  position: relative;
  font-family: 'Boogaloo';
  font-size: 2em;
  text-shadow: 1px 1px 4px #080808;
  background-color: #181818ba;
  padding: 4px 20px;
  animation: 0.7s ease-out 0s 1 slideInFromTop;
}

.join-team-message {
  background-color: #181818d1;
  padding: 20px 20px;
  margin-bottom: 20px;
}

.join-team-message h1 {
  font-family: 'Boogaloo';
  font-size: 2.6em;
  text-shadow: 1px 1px 4px #080808;
  padding-bottom: 20px;
}

.join-team-message p {
  font-size: 1.2em;
  line-height: normal;
  letter-spacing: 1px;
  text-shadow: 2px 2px 1px #181818d1, -2px -2px 1px #181818d1;
  /* text-shadow: none; */
}

.join-container {
    padding: 20px 0;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    text-shadow: 1px 1px 3px black;
    z-index: 2;
}

.g-img-left {
    /* height: 300px;  */
    bottom: -60px;
    left: -50px;
    position: fixed;
    overflow: hidden;
    transform: rotate(0.06turn);
    
    animation: 1s ease-out 0s 1 slideInFromLeft;
}
.g-img-left img{
  max-height: 35vh;
}
@keyframes slideInFromLeft {
    0% {
      transform: translateX(-100%) rotate(0.06turn);
    }
    100% {
      transform: translateX(0) rotate(0.06turn);
    }
}
.g-img-right {
    bottom: -5px;
    right: -20px;
    position: fixed;
    overflow: hidden;
    animation-delay: 2s;
    animation: 1s ease-out 0s 1 slideInFromRight;
}
.g-img-right img{
    max-height: 35vh;
    max-width: 60vw;
}

@keyframes slideInFromRight {
    0% {
      transform: translateX(100%);
    }
    100% {
      transform: translateX(0);
    }
}