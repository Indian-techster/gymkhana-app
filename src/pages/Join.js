import React from 'react'
import { connect } from 'react-redux'
import { restartPoints } from '../js/actions/index'
import JoinTeam from '../components/JoinTeam'
import JoinUser from '../components/JoinUser'
import "./styles/Join.css"

/** Redux function. Sirve para enviar (dispatch) acciones al store */
function mapDispatchToProps(dispatch) {
    return {
        restartPoints: points => dispatch(restartPoints())
    }
}

const mapStateToProps = state => {
    return {
        game: state.game,
        gameInfo: state.gameInfo,
        username: state.username,
    }
}

class Inicio extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            step: 1,
        }
        const gameData = this.props.gameInfo ? this.props.gameInfo[0] : null
        this.welcomeMessage = gameData ? gameData.welcomeMessage : "Welcome"
        this.joinTeamMessage = gameData ? gameData.joinTeamMessage : ""
        this.image1 = gameData && gameData.joinPage ? this.tryRequire(gameData.joinPage[0]) : null
        this.image2 = gameData && gameData.joinPage ? this.tryRequire(gameData.joinPage[1]) : null
     }

    componentWillMount() {
        if(this.props.username)  {
            this.setState({step: 2})
        }
    }

    tryRequire(image) {
        try {
         return require('../images/'+image);
        } catch (err) {
         return null;
        }
    }

    render() {
        return <React.Fragment>
            <div className="join-container">
                {!this.props.username &&
                    <JoinUser welcomeMessage={this.welcomeMessage}/>
                }
                {this.props.username &&
                    <JoinTeam message={this.joinTeamMessage}/>
                }
            </div>
            <div>
            {this.image1 && <div className="join-img-left"><img src={this.image1} alt={this.image1}/></div>}
            {this.image2 && <div className="g-img-right"><img src={this.image2} alt={this.image2}/></div>}
            </div>

        </React.Fragment>
    }
}

const inicioConnected = connect(mapStateToProps, mapDispatchToProps)(Inicio)
export default inicioConnected;
