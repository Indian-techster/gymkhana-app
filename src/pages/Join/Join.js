import React from 'react'
import { connect } from 'react-redux'
import { Redirect } from 'react-router-dom'
import JoinTeam from './components/JoinTeam'
import JoinUser from './components/JoinUser'
import "./Join.css"
import { setUsername, setUserId, setTeam, setTeamId } from 'js/actions/index'

/** Redux function. Sirve para enviar (dispatch) acciones al store */
function mapDispatchToProps(dispatch) {
    return {
        setUserId: userid => dispatch(setUserId(userid)),
        setUsername: username => dispatch(setUsername(username)),
        setTeamId: teamId => dispatch(setTeamId(teamId)),
        setTeam: team => dispatch(setTeam(team)), //TODO cambiar a teamname
    }
}
const mapStateToProps = state => {
    return {
        gameid: state.game,
        gameInfo: state.gameInfo,
        userid: state.userid,
        username: state.username,
        team: state.team
    }
}

export class Join extends React.Component {

    constructor(props) {
        super(props)
        this.setUser = this.setUser.bind(this)
        this.setTeam = this.setTeam.bind(this)
        const gameData = this.props.gameInfo ? this.props.gameInfo[0] : null
        this.welcomeMessage = gameData ? gameData.welcomeMessage : "Welcome"
        this.joinTeamMessage = gameData ? gameData.joinTeamMessage : ""
        this.image1 = gameData && gameData.joinPage ? this.tryRequire(gameData.joinPage[0]) : null
        this.image2 = gameData && gameData.joinPage ? this.tryRequire(gameData.joinPage[1]) : null
     }

    tryRequire(image) {
        try {
         return require('images/'+image);
        } catch (err) {
         return null;
        }
    }

    setUser(userid, username) {
        this.props.setUserId(userid)
        this.props.setUsername(username) //esto provocará que se actualice el state y se renderice la página con joinTeam
    }

    setTeam(teamid, teamname) {
        this.props.setTeamId(teamid)
        this.props.setTeam(teamname)
    }

    render() {
        const userSetted = (this.props.userid && this.props.username) ? true : false
        if(userSetted && this.props.team) {
            return <Redirect to='/intro' />
        }
        return (
            <React.Fragment>
                <div className="join-container">
                    {!userSetted && 
                        <JoinUser   welcomeMessage={this.welcomeMessage} 
                                    gameid={this.props.gameid} 
                                    setUser={this.setUser}/>  }
                    {userSetted && 
                        <JoinTeam   message = {this.joinTeamMessage} 
                                    gameid = {this.props.gameid} 
                                    username = {this.props.username}
                                    userid = {this.props.userid}
                                    setTeam = {this.setTeam}/> }
                </div>
                <div>
                    {this.image1 && <div className="join-img-left"><img src={this.image1} alt={this.image1}/></div>}
                    {this.image2 && <div className="g-img-right"><img src={this.image2} alt={this.image2}/></div>}
                </div>
            </React.Fragment>
        )
    }
}

const inicioConnected = connect(mapStateToProps, mapDispatchToProps)(Join)
export default inicioConnected;
