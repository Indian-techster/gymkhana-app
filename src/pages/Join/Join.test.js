import React from 'react'
import { shallow } from 'enzyme'
import { Join } from './Join'
import JoinUser from './components/JoinUser'
import JoinTeam from './components/JoinTeam'
import { render } from '@testing-library/react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'

const renderWithRouterrender = (
    ui,
    {
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    } = {}
) => {
    return {
        ...render(<Router history={history}>{ui}</Router>),
        history,
    }
}

let component

beforeEach(() => {
    component = shallow(<Join/>);
})

describe('Join Page', () => {
    it('renders correctly', () => {
        expect(component).toMatchSnapshot()
    })

    it('displays joinUser Component when username is not defined', () => {
        const component = shallow(<Join/>)
        expect(component.find(JoinUser).length).toBe(1)
        expect(component.find(JoinTeam).length).toBe(0)
    })
    
    it('displays joinTeam Component when username is defined but team is not', () => {
        const component = shallow(<Join userid='1'/>)
        expect(component.find(JoinUser).length).toBe(0)
        expect(component.find(JoinTeam).length).toBe(1)
    })

    it('redirects to intro page when username and team are defined', async () => {
        const { history } = renderWithRouterrender(<Join userid='1' team='1'/>)
        await expect(history.location.pathname).toEqual('/intro')
    })
})