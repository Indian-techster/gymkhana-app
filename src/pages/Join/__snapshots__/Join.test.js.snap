// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Join Page renders correctly 1`] = `
<Fragment>
  <div
    className="join-container"
  >
    <JoinUser
      setUser={[Function]}
      welcomeMessage="Welcome"
    />
  </div>
  <div />
</Fragment>
`;
